### DISCUSSION TITLE
2D segmentation runtime error

### DISCUSSION THREAD

**Discussion by witness97** [Project-MONAI/MONAI#2302] [Q&A]
Hi MONAI team and thanks for providing such amazing package. I try to run 2D_segmentation Unet training code on my own dataset, but it got the error

RuntimeError: Caught RuntimeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/monai/transforms/transform.py", line 48, in apply_transform
    return transform(data)
  File "/usr/local/lib/python3.7/dist-packages/monai/transforms/croppad/array.py", line 327, in __call__
    self.randomize(img.shape[1:])
  File "/usr/local/lib/python3.7/dist-packages/monai/transforms/croppad/array.py", line 315, in randomize
    self._size = fall_back_tuple(self.roi_size, img_size)
  File "/usr/local/lib/python3.7/dist-packages/monai/utils/misc.py", line 174, in fall_back_tuple
    user = ensure_tuple_rep(user_provided, ndim)
  File "/usr/local/lib/python3.7/dist-packages/monai/utils/misc.py", line 131, in ensure_tuple_rep
    raise ValueError(f"Sequence must have length {dim}, got {len(tup)}.")
ValueError: Sequence must have length 3, got 2.

The shape of image in my dataset is (512, 512, 3), I just wonder why this error would happen and how to turn my data to fit the code. Thanks.

**Comment by Nic-Ma** [comment]
Hi @witness97 ,

Thanks for your interest here.
The difference between the example data and your data is that: your data is `channel-last` and has 3 channels, the example data has no `channel` dim.
So you can change the `AddChanneld` transform to `EnsureChannelFirstd` transform and try again.

Thanks.

**Reply by witness97** [reply]
Hi @Nic-Ma ,

Really thanks for your reply. I still have two questions. When I change AddChanneld transform to EnsureChannelFirstd transform, what value of key I should set in EnsureChannelFirstd function? Moreover, the in_channels = 1 in the training Unet but my images have 3 channels, I just wonder if it is OK for the train-net to put all 3 channels in? Will this cause the mean dice going wrong? Hope to get your reply.

Thanks

**Reply by Nic-Ma** [reply]
Hi @witness97 ,

You should put both your `image` and `label` into `EnsureChannelFirstd` transform, it will automatically adjust the shape to ensure channel dim first.
About the input channel of UNet, please change the parameter of UNet to match it.
About the performance, I think that depends on many things, need to do experiments and tune it.

Thanks.

**Reply by witness97** [reply]
Hi @Nic-Ma ,
I try to put my images and segs like this

train_imtrans = Compose(
        [
            LoadImage(image_only=True),
            ScaleIntensity(),
            EnsureChannelFirstd(keys= images),
         
            RandSpatialCrop((96, 96), random_size=False),
            RandRotate90(prob=0.5, spatial_axes=(0, 1)),
            ToTensor(),
        ]
    )
    train_segtrans = Compose(
        [
            LoadImage(image_only=True),
            EnsureChannelFirstd(keys= segs),
            RandSpatialCrop((96, 96), random_size=False),
            RandRotate90(prob=0.5, spatial_axes=(0, 1)),
            ToTensor(),
        ]
    )

and I turn all the parameters of Unet to fit it. However, I still meet the problem: 
'dictionary update sequence element #0 has length 512; 2 is required',
I guess it may still caused by the format error of the training set. I just wonder how EnsureChannelFirstd works and why this problem would happen? 
Thanks!

**Reply by Nic-Ma** [reply]
Hi @witness97 ,

We have 2 sets of transforms: dictionary-based and array-based.
The name of dictionary-based transforms ends with a "d", like `EnsureChannelFirstd`.
You should not compose the dictionary-based transforms and array-based transforms together.
I would suggest to make your data as a dictionary and use all dictionary-based transforms.
You can refer to: https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d.ipynb

Thanks.

**Reply by witness97** [reply]
Hi @Nic-Ma ,

Really thanks for your quick reply which helps me a lot. I can successfully run the 2D segmentation in dictionary-based version. However, I face the same problem as Q&A #1455 that the mean dice I get from the Unet is always larger than 1. I tried the approach in #1455 to turn the crop areas but it seems not work well. My dataset are images with multi-classes and size (512,512,3), I just wonder if this incorrect mean dice is caused by the dataset or something else? 

Thanks.

**Reply by Nic-Ma** [reply]
Hi @rijobro , according to your training experience, what do you think the possible reason that mean dice > 1?

Thanks.

**Reply by rijobro** [reply]
Hi @witness97 , I'll try to put together a multi-class segmentation tutorial, I think this will be useful for quite a few users. I'll let you know once it's done. see here: https://github.com/Project-MONAI/tutorials/issues/247.

**Reply by rijobro** [reply]
Sorry, as @Nic-Ma pointed out, this the [brats segmentation tutorial](https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/brats_segmentation_3d.ipynb) is already multi-class. 
I would suggest checking where your code differs from that tutorial, especially in the post-processing, and the calculation of the Dice score.

If you still can't get it working, feel free to send a minimum working example (using a public dataset), and we can have a look.
